pkg/southbound/synchronizer/sessionManager.go

```diff
@@ -0,0 +1,249 @@
+	return func(sessionManager *SessionManager) {
+		sessionManager.operationalStateCache = operationalStateCache
+	}
+}
+
+// WithNewTargetFn sets southbound target function
+func WithNewTargetFn(newTargetFn func() southbound.TargetIf) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.newTargetFn = newTargetFn
+	}
+}
+
+// WithOperationalStateCacheLock sets operational state cache lock
+func WithOperationalStateCacheLock(operationalStateCacheLock *sync.RWMutex) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.operationalStateCacheLock = operationalStateCacheLock
+	}
+}
+
+// WithDeviceChangeStore sets device change store
+func WithDeviceChangeStore(deviceChangeStore device.Store) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.deviceChangeStore = deviceChangeStore
+	}
+}
+
+// Start starts session manager
+func (sm *SessionManager) Start() error {
+	log.Info("Session manager started")
+	err := sm.deviceStore.Watch(sm.topoChannel)
+	if err != nil {
+		return err
+	}
+
+	go sm.processDeviceEvents(sm.topoChannel)
+	return nil
+}
+
+// processDeviceEvents process incoming device events
+func (sm *SessionManager) processDeviceEvents(ch <-chan *topodevice.ListResponse) {
+	for event := range ch {
+		log.Infof("Received event type %v for device %v", event.Type, event.Device)
+		err := sm.processDeviceEvent(event)
+		if err != nil {
+			log.Errorf("Error updating session %v", event.Device.ID, err)
+		}
+
+	}
+}
+
+// processDeviceEvent process a device event
+func (sm *SessionManager) processDeviceEvent(event *topodevice.ListResponse) error {
+
+	switch event.Type {
+	case topodevice.ListResponse_ADDED:
+		log.Info("Process device event Added")
+		err := sm.createSession(event.Device)
+		if err != nil {
+			return err
+		}
+
+	case topodevice.ListResponse_NONE:
+		err := sm.createSession(event.Device)
+		if err != nil {
+			return err
+		}
+	case topodevice.ListResponse_UPDATED:
+		log.Info("Process device event updated")
+
+	case topodevice.ListResponse_REMOVED:
+		err := sm.deleteSession(event.Device)
+		if err != nil {
+			return err
+		}
+
+	}
+	return nil
+
+}
+
+// createSession creates a new gNMI session
+func (sm *SessionManager) createSession(device *topodevice.Device) error {
+	log.Info("Creating session for device:", device.ID)
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+
+	errChan := make(chan events.DeviceResponse)
+	session := &Session{
+		opStateChan:               sm.opStateChan,
+		southboundErrorChan:       errChan,
+		dispatcher:                sm.dispatcher,
+		modelRegistry:             sm.modelRegistry,
+		operationalStateCache:     sm.operationalStateCache,
+		operationalStateCacheLock: sm.operationalStateCacheLock,
+		deviceChangeStore:         sm.deviceChangeStore,
+		device:                    device,
+		target:                    sm.newTargetFn(),
```

Comment:
I tried to avoid any code changes which are not related to session management thing but these are minor things and I will improve them in the same PR.